File: cmd/edit.go
package cmd

import (
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"os/exec"

	"github.com/benoctopus/sesh/internal/config"
	"github.com/rotisserie/eris"
	"github.com/spf13/cobra"
)

// editCmd represents the edit command
var editCmd = &cobra.Command{
	Use:   "edit",
	Short: "Edit the sesh configuration file",
	Long: `Opens the sesh configuration file in your default editor.

The config file is located at:
  - Linux: ~/.config/sesh/config.yaml
  - macOS: ~/Library/Application Support/sesh/config.yaml
  - Windows: %APPDATA%\sesh\config.yaml

The editor is determined by the EDITOR environment variable (falls back to vim).

After editing, the configuration will be validated. If validation fails, you'll
be prompted to fix the errors before the changes are saved.

Example config.yaml:
  version: "1"
  workspace_dir: ~/projects
  session_backend: tmux
  fuzzy_finder: fzf
  startup_command: ""
`,
	RunE: runEdit,
}

func init() {
	rootCmd.AddCommand(editCmd)
}

func runEdit(cmd *cobra.Command, args []string) error {
	// Get config path
	configPath, err := config.GetConfigPath()
	if err != nil {
		return eris.Wrap(err, "failed to get config path")
	}

	// Ensure config directory exists
	if err := config.EnsureConfigDir(); err != nil {
		return eris.Wrap(err, "failed to ensure config directory")
	}

	// Create config file with defaults if it doesn't exist
	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		if err := createDefaultConfig(configPath); err != nil {
			return eris.Wrap(err, "failed to create default config")
		}
		fmt.Printf("Created default config at: %s\n", configPath)
	}

	// Get the file hash before editing
	hashBefore, err := hashFile(configPath)
	if err != nil {
		return eris.Wrap(err, "failed to hash config file")
	}

	// Open in editor
	editor := getEditor()
	editorCmd := exec.Command(editor, configPath)
	editorCmd.Stdin = os.Stdin
	editorCmd.Stdout = os.Stdout
	editorCmd.Stderr = os.Stderr

	if err := editorCmd.Run(); err != nil {
		return eris.Wrapf(err, "failed to run editor: %s", editor)
	}

	// Get the file hash after editing
	hashAfter, err := hashFile(configPath)
	if err != nil {
		return eris.Wrap(err, "failed to hash config file after editing")
	}

	// Check if file was modified
	if hashBefore == hashAfter {
		fmt.Println("No changes made to config")
		return nil
	}

	// Validate the config
	if err := config.ValidateConfigFile(configPath); err != nil {
		fmt.Printf("\nConfig validation failed: %+v\n", eris.ToString(err, false))
		fmt.Printf("\nThe config file has errors. Do you want to:\n")
		fmt.Printf("  1. Edit again to fix errors\n")
		fmt.Printf("  2. Discard changes and restore previous version\n")
		fmt.Printf("  3. Save anyway (not recommended)\n")
		fmt.Printf("\nChoice (1-3): ")

		var choice string
		fmt.Scanln(&choice)

		switch choice {
		case "1":
			// Edit again
			return runEdit(cmd, args)
		case "2":
			// We can't easily restore without keeping a backup
			// So we'll just tell the user to manually fix it
			fmt.Println("\nPlease manually fix the config file or delete it to start over.")
			return eris.New("config validation failed")
		case "3":
			// Save anyway
			fmt.Println("\nWarning: Saving invalid config. This may cause issues.")
			return nil
		default:
			return eris.New("invalid choice")
		}
	}

	fmt.Printf("Config saved and validated successfully: %s\n", configPath)
	return nil
}

// getEditor returns the user's preferred editor
// Priority: VISUAL > EDITOR > vi
func getEditor() string {
	if editor := os.Getenv("VISUAL"); editor != "" {
		return editor
	}
	if editor := os.Getenv("EDITOR"); editor != "" {
		return editor
	}
	return "vi"
}

// createDefaultConfig creates a default configuration file
func createDefaultConfig(configPath string) error {
	// Load default config
	cfg := &config.Config{
		WorkspaceDir:   "~/.sesh",
		SessionBackend: "auto",
		StartupCommand: "",
		FuzzyFinder:    "auto",
	}

	return config.SaveConfig(cfg)
}

// hashFile computes the SHA256 hash of a file
func hashFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

File: src/cli/edit.rs
use crate::config;
use crate::error::{Error, Result};
use anyhow::Context;
use sha2::{Digest, Sha256};
use std::io::{self, Write};
use std::process::Command;

/// Get the user's preferred editor from environment variables
/// Priority: VISUAL > EDITOR > vi
fn get_editor() -> String {
    std::env::var("VISUAL")
        .or_else(|_| std::env::var("EDITOR"))
        .unwrap_or_else(|_| "vi".to_string())
}

/// Compute SHA256 hash of a file
fn hash_file(path: &std::path::Path) -> anyhow::Result<String> {
    let content = std::fs::read(path)
        .with_context(|| format!("Failed to read file for hashing: {}", path.display()))?;

    let mut hasher = Sha256::new();
    hasher.update(&content);
    let result = hasher.finalize();

    Ok(hex::encode(result))
}

/// Prompt user for a choice when validation fails
fn prompt_validation_failure() -> anyhow::Result<String> {
    println!("\nThe config file has errors. What would you like to do?");
    println!("  1. Edit again to fix errors");
    println!("  2. Discard changes (config will remain invalid until fixed)");
    println!("  3. Keep invalid config anyway (not recommended)");
    print!("\nChoice (1-3): ");
    io::stdout().flush()?;

    let mut input = String::new();
    io::stdin().read_line(&mut input)?;

    Ok(input.trim().to_string())
}

/// Run the edit command
pub async fn run() -> Result<()> {
    // Ensure config directory exists
    config::ensure_config_dir().map_err(|e| Error::ConfigError(e.to_string()))?;

    let config_path = config::config_path().map_err(|e| Error::ConfigError(e.to_string()))?;

    // Create config with defaults if it doesn't exist
    if !config_path.exists() {
        config::save_default().map_err(|e| Error::ConfigError(e.to_string()))?;
        println!("Created default config at: {}", config_path.display());
    }

    // Compute hash before editing
    let hash_before =
        hash_file(&config_path).context("Failed to hash config file before editing")?;

    // Open in editor
    let editor = get_editor();
    let status = Command::new(&editor)
        .arg(&config_path)
        .status()
        .with_context(|| format!("Failed to run editor: {}", editor))?;

    if !status.success() {
        return Err(Error::ConfigError(format!(
            "Editor exited with error status: {}",
            status
        )));
    }

    // Compute hash after editing
    let hash_after = hash_file(&config_path).context("Failed to hash config file after editing")?;

    // Check if file was modified
    if hash_before == hash_after {
        println!("No changes made to config");
        return Ok(());
    }

    // Validate the config
    loop {
        match config::load() {
            Ok(cfg) => {
                // Config parsed successfully, now validate values
                if let Err(e) = cfg.validate() {
                    eprintln!("\nConfig validation failed: {}", e);

                    // Check if we're in an interactive terminal
                    if !std::io::IsTerminal::is_terminal(&std::io::stdin()) {
                        return Err(Error::ConfigError(format!(
                            "Config validation failed in non-interactive mode. Please fix manually: {}",
                            config_path.display()
                        )));
                    }

                    let choice = prompt_validation_failure()?;

                    match choice.as_str() {
                        "1" => {
                            // Edit again
                            let status = Command::new(&editor)
                                .arg(&config_path)
                                .status()
                                .with_context(|| format!("Failed to run editor: {}", editor))?;

                            if !status.success() {
                                return Err(Error::ConfigError(format!(
                                    "Editor exited with error status: {}",
                                    status
                                )));
                            }
                            // Continue the loop to validate again
                            continue;
                        }
                        "2" => {
                            // Discard changes - user needs to manually fix
                            println!(
                                "\nPlease manually fix the config file or delete it to start over."
                            );
                            return Err(Error::ConfigError("Config validation failed".to_string()));
                        }
                        "3" => {
                            // Save anyway
                            println!(
                                "\nWarning: Config file contains errors. This may cause issues."
                            );
                            return Ok(());
                        }
                        _ => {
                            return Err(Error::ConfigError("Invalid choice".to_string()));
                        }
                    }
                } else {
                    // Validation succeeded
                    break;
                }
            }
            Err(e) => {
                eprintln!("\nConfig parsing failed: {}", e);

                // Check if we're in an interactive terminal
                if !std::io::IsTerminal::is_terminal(&std::io::stdin()) {
                    return Err(Error::ConfigError(format!(
                        "Config parsing failed in non-interactive mode. Please fix manually: {}",
                        config_path.display()
                    )));
                }

                let choice = prompt_validation_failure()?;

                match choice.as_str() {
                    "1" => {
                        // Edit again
                        let status = Command::new(&editor)
                            .arg(&config_path)
                            .status()
                            .with_context(|| format!("Failed to run editor: {}", editor))?;

                        if !status.success() {
                            return Err(Error::ConfigError(format!(
                                "Editor exited with error status: {}",
                                status
                            )));
                        }
                        // Continue the loop to validate again
                        continue;
                    }
                    "2" => {
                        // Discard changes - user needs to manually fix
                        println!(
                            "\nPlease manually fix the config file or delete it to start over."
                        );
                        return Err(Error::ConfigError("Config parsing failed".to_string()));
                    }
                    "3" => {
                        // Save anyway
                        println!("\nWarning: Config file contains errors. This may cause issues.");
                        return Ok(());
                    }
                    _ => {
                        return Err(Error::ConfigError("Invalid choice".to_string()));
                    }
                }
            }
        }
    }

    println!(
        "Config saved and validated successfully: {}",
        config_path.display()
    );
    Ok(())
}
